websocket/connection.go

```diff
@@ -137,90 +151,115 @@ func newConnection(target string, messageChan chan []byte) *ManagedConnection {
+	var err error
 	wait.ExponentialBackoff(c.connectionBackoff, func() (bool, error) {
-		var conn rawConnection
-		conn, err = connFactory(c.target)
-		if err != nil {
-			return false, nil
-		}
-		c.connectionLock.Lock()
-		defer c.connectionLock.Unlock()
+		select {
+		default:
+			var conn rawConnection
+			conn, err = c.connectionFactory()
+			if err != nil {
+				return false, nil
+			}
+
+			c.connectionLock.Lock()
+			defer c.connectionLock.Unlock()
 
-		c.connection = conn
-		return true, nil
+			c.connection = conn
+			return true, nil
+		case <-c.closeChan:
+			err = errShuttingDown
+			return false, err
+		}
 	})
 
 	return err
 }
 
-// keepalive keeps the connection open and reads control messages.
-// All messages are discarded.
-func (c *ManagedConnection) keepalive() (err error) {
+// keepalive keeps the connection open.
+func (c *ManagedConnection) keepalive() error {
+	for {
+		select {
+		default:
+			if err := c.read(); err != nil {
+				return err
+			}
+		case <-c.closeChan:
+			return errShuttingDown
+		}
+	}
+}
+
+// closeConnection closes the underlying websocket connection.
+func (c *ManagedConnection) closeConnection() error {
+	c.connectionLock.Lock()
+	defer c.connectionLock.Unlock()
+
+	if c.connection != nil {
+		err := c.connection.Close()
+		c.connection = nil
+		return err
+	}
+	return nil
+}
+
+// read reads the next message from the connection.
+// If a messageChan is supplied and the current message type is not
+// a control message, the message is sent to that channel.
+func (c *ManagedConnection) read() error {
+	c.connectionLock.RLock()
+	defer c.connectionLock.RUnlock()
+
+	if c.connection == nil {
+		return ErrConnectionNotEstablished
+	}
+
 	c.readerLock.Lock()
 	defer c.readerLock.Unlock()
 
-	for {
-		func() {
-			c.connectionLock.RLock()
-			defer c.connectionLock.RUnlock()
-
-			if conn := c.connection; conn != nil {
-				var reader io.Reader
-				var messageType int
-				messageType, reader, err = conn.NextReader()
-				if err != nil {
-					conn.Close()
-				}
-
-				// Send the message to the channel if its an application level message
-				// and if that channel is set.
-				if c.messageChan != nil && (messageType == websocket.TextMessage || messageType == websocket.BinaryMessage) {
-					if message, _ := ioutil.ReadAll(reader); message != nil {
-						c.messageChan <- message
-					}
-				}
-			} else {
-				err = ErrConnectionNotEstablished
-			}
-		}()
+	messageType, reader, err := c.connection.NextReader()
+	if err != nil {
+		return err
+	}
 
-		if err != nil {
-			return err
+	// Send the message to the channel if its an application level message
+	// and if that channel is set.
+	// TODO(markusthoemmes): Return the messageType along with the payload.
+	if c.messageChan != nil && (messageType == websocket.TextMessage || messageType == websocket.BinaryMessage) {
+		if message, _ := ioutil.ReadAll(reader); message != nil {
+			c.messageChan <- message
```

Comment:
Orthogonal: do we care about sending/receiving Ping/Pong messages to actually keep the connection alive?

Reply:
Oh I haven't thought about this. This might be very helpful especially for the websocket connection of the activator which might go quite a while without ever sending a message. Great point, I'll open an issue to track that addition.